// @sitecore-marketplace-sdk/core/dist/index.cjs.js
'use strict';

/**
 * Enumeration of all possible error codes that can be thrown by the SDK.
 * These codes help identify the type and source of errors for better error handling.
 */
exports.ErrorCode = void 0;
(function (ErrorCode) {
    // Communication Errors
    ErrorCode["TIMEOUT"] = "TIMEOUT";
    ErrorCode["CONNECTION_ERROR"] = "CONNECTION_ERROR";
    ErrorCode["HANDSHAKE_FAILED"] = "HANDSHAKE_FAILED";
    ErrorCode["INVALID_ORIGIN"] = "INVALID_ORIGIN";
    ErrorCode["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
    // Request/Response Errors
    ErrorCode["INVALID_REQUEST"] = "INVALID_REQUEST";
    ErrorCode["METHOD_NOT_FOUND"] = "METHOD_NOT_FOUND";
    ErrorCode["EXECUTION_ERROR"] = "EXECUTION_ERROR";
    // State/Event Errors
    ErrorCode["SUBSCRIPTION_ERROR"] = "SUBSCRIPTION_ERROR";
    ErrorCode["INVALID_STATE"] = "INVALID_STATE";
    // Runtime Errors
    ErrorCode["NOT_INITIALIZED"] = "NOT_INITIALIZED";
    ErrorCode["ALREADY_INITIALIZED"] = "ALREADY_INITIALIZED";
    ErrorCode["NOT_CONNECTED"] = "NOT_CONNECTED";
    ErrorCode["INITIALIZATION_ERROR"] = "INITIALIZATION_ERROR";
    // Host Specific
    ErrorCode["HOST_NOT_READY"] = "HOST_NOT_READY";
    ErrorCode["TOKEN_ERROR"] = "TOKEN_ERROR";
    // Client Specific
    ErrorCode["CLIENT_NOT_READY"] = "CLIENT_NOT_READY";
    ErrorCode["IFRAME_ERROR"] = "IFRAME_ERROR";
})(exports.ErrorCode || (exports.ErrorCode = {}));
/**
 * Core error class that implements the SDKError interface.
 * Provides static factory methods for creating common error types.
 *
 * @example
 * ```typescript
 * // Creating a timeout error
 * throw CoreError.timeout();
 *
 * // Creating a method not found error
 * throw CoreError.methodNotFound('getUserInfo');
 * ```
 */
class CoreError extends Error {
    code;
    details;
    constructor(code, message, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'CoreError';
    }
    /**
     * Creates a timeout error.
     * @param details - Optional details about the timeout
     */
    static timeout(details) {
        return new CoreError(exports.ErrorCode.TIMEOUT, 'Request timed out', details);
    }
    /**
     * Creates a handshake failed error.
     * @param details - Optional details about the failure
     */
    static handshakeFailed(details) {
        return new CoreError(exports.ErrorCode.HANDSHAKE_FAILED, 'Failed to establish connection', details);
    }
    /**
     * Creates an invalid origin error.
     * @param origin - The invalid origin that was detected
     */
    static invalidOrigin(origin) {
        return new CoreError(exports.ErrorCode.INVALID_ORIGIN, `Invalid message origin: ${origin}`, { origin });
    }
    /**
     * Creates a not initialized error.
     */
    static notInitialized() {
        return new CoreError(exports.ErrorCode.NOT_INITIALIZED, 'SDK not properly initialized');
    }
    /**
     * Creates an already initialized error.
     */
    static alreadyInitialized() {
        return new CoreError(exports.ErrorCode.ALREADY_INITIALIZED, 'SDK is already initialized');
    }
    /**
     * Creates a method not found error.
     * @param method - The name of the method that wasn't found
     */
    static methodNotFound(method) {
        return new CoreError(exports.ErrorCode.METHOD_NOT_FOUND, `Method not found: ${method}`, { method });
    }
    /**
     * Creates an execution error.
     * @param error - The underlying error that caused the execution failure
     */
    static executionError(error) {
        return new CoreError(exports.ErrorCode.EXECUTION_ERROR, 'Method execution failed', error);
    }
}

class GenericResponseData {
    status;
    statusText;
    headers;
    body;
    constructor(status, statusText, headers, body) {
        this.status = status;
        this.statusText = statusText;
        this.headers = headers;
        this.body = body;
    }
    /**
     * Parses the body of the response as JSON.
     * @returns {Promise<any>} The parsed JSON object.
     */
    async json() {
        return JSON.parse(await this.text());
    }
    /**
     * Parses the body of the response as a string.
     * @returns {Promise<string>} The parsed text.
     */
    text() {
        return Promise.resolve(new TextDecoder().decode(this.body));
    }
    /**
     * Returns the body as an ArrayBuffer.
     * @returns {Promise<ArrayBuffer>} The raw binary data.
     */
    arrayBuffer() {
        return Promise.resolve(this.body);
    }
    /**
     * Returns the body as a Blob.
     * @returns {Promise<Blob>} The Blob representation of the response body.
     */
    blob() {
        return Promise.resolve(new Blob([this.body], {
            type: this.headers['content-type'] || 'application/octet-stream',
        }));
    }
}

/**
 * @deprecated Use AllowedExtensionPoints instead
 */
exports.AllowedTouchpoints = void 0;
(function (AllowedTouchpoints) {
    AllowedTouchpoints["standalone"] = "standalone";
    AllowedTouchpoints["xmcFullscreen"] = "xmc:fullscreen";
    AllowedTouchpoints["xmcPagesContextPanel"] = "xmc:pages:contextpanel";
    AllowedTouchpoints["xmcPagesCustomField"] = "xmc:pages:customfield";
    AllowedTouchpoints["xmcDashboardBlocks"] = "xmc:dashboardblocks";
})(exports.AllowedTouchpoints || (exports.AllowedTouchpoints = {}));
/**
 * Allowed extension points for Sitecore Marketplace applications.
 * This enum defines the valid extension points where applications can be embedded.
 */
exports.AllowedExtensionPoints = void 0;
(function (AllowedExtensionPoints) {
    AllowedExtensionPoints["standalone"] = "standalone";
    AllowedExtensionPoints["xmcFullscreen"] = "xmc:fullscreen";
    AllowedExtensionPoints["xmcPagesContextPanel"] = "xmc:pages:contextpanel";
    AllowedExtensionPoints["xmcPagesCustomField"] = "xmc:pages:customfield";
    AllowedExtensionPoints["xmcDashboardBlocks"] = "xmc:dashboardblocks";
})(exports.AllowedExtensionPoints || (exports.AllowedExtensionPoints = {}));

/**
 * A simple event emitter implementation that provides type-safe event handling.
 * Supports subscribing to events, unsubscribing, and emitting events with payloads.
 * Events are handled synchronously and any errors in handlers are caught and logged.
 */
class EventEmitter {
    /** Map of event names to their set of handlers */
    handlers = new Map();
    /**
     * Subscribes to an event.
     * @param event - The name of the event to listen for
     * @param handler - The callback function to execute when the event occurs
     * @returns A function that can be called to unsubscribe from the event
     *
     * @example
     * ```typescript
     * const unsubscribe = emitter.on('userUpdate', (data: UserData) => {
     *   console.log('User updated:', data);
     * });
     *
     * // Later, to stop listening:
     * unsubscribe();
     * ```
     */
    on(event, handler) {
        if (!this.handlers.has(event)) {
            this.handlers.set(event, new Set());
        }
        this.handlers.get(event).add(handler);
        // Return unsubscribe function
        return () => this.off(event, handler);
    }
    /**
     * Unsubscribes a handler from an event.
     * If this was the last handler for the event, the event is removed from the handlers map.
     * @param event - The name of the event to unsubscribe from
     * @param handler - The handler function to remove
     */
    off(event, handler) {
        this.handlers.get(event)?.delete(handler);
        // Clean up empty sets
        if (this.handlers.get(event)?.size === 0) {
            this.handlers.delete(event);
        }
    }
    /**
     * Emits an event with the provided data. In addition to calling handlers for the specific event,
     * it also calls any wildcard ('*') handlers. Wildcard handlers receive an object with { event, payload }.
     */
    emit(event, data) {
        // Emit to specific handlers.
        this.handlers.get(event)?.forEach((handler) => {
            try {
                handler(data);
            }
            catch (error) {
                console.error('Error in event handler for %s:', event, error);
            }
        });
        // Emit to wildcard handlers.
        this.handlers.get('*')?.forEach((handler) => {
            try {
                // Wildcard handler signature: (data: { event: string; payload: T }) => void
                handler({ event, payload: data });
            }
            catch (error) {
                console.error('Error in wildcard handler for %s:', event, error);
            }
        });
    }
    /**
     * Removes all event handlers.
     * After calling this method, no events will be handled until new handlers are registered.
     */
    clear() {
        this.handlers.clear();
    }
    /**
     * Checks if an event has any active listeners.
     * @param event - The name of the event to check
     * @returns True if the event has at least one handler, false otherwise
     */
    hasListeners(event) {
        return this.handlers.has(event) && this.handlers.get(event).size > 0;
    }
}

const HANDSHAKE_INIT_EVENT = 'handshake:init';
const HANDSHAKE_RESPONSE_EVENT = 'handshake:response';
const RETRY_DELAY = 2000; // 2 seconds between retries
const MAX_RETRIES = 5;
/**
 * HandshakeManager handles the handshake process between host and client applications.
 * It abstracts the complexity of establishing a secure connection between the two sides.
 */
class HandshakeManager {
    config;
    eventEmitter = new EventEmitter();
    handshakeComplete = false;
    retryCount = 0;
    retryTimeout = null;
    pendingInitMessage = null;
    constructor(config) {
        this.config = config;
    }
    /**
     * Initializes the handshake process for the client side.
     * This method sends a handshake initialization message and waits for a response.
     *
     * @returns A promise that resolves when the handshake is complete.
     */
    async initializeClient(sendMessage) {
        if (this.handshakeComplete) {
            return;
        }
        return new Promise((resolve, reject) => {
            const initTimeout = setTimeout(() => {
                if (!this.handshakeComplete && this.retryCount < MAX_RETRIES) {
                    this.retryCount++;
                    console.debug(`[client SDK] Handshake timeout, retrying (${this.retryCount}/${MAX_RETRIES})...`);
                    this.retryTimeout = setTimeout(() => {
                        this.initializeClient(sendMessage).then(resolve).catch(reject);
                    }, RETRY_DELAY);
                }
                else if (!this.handshakeComplete) {
                    reject(new CoreError(exports.ErrorCode.TIMEOUT, '[client SDK] Handshake timed out after maximum retries'));
                }
            }, RETRY_DELAY);
            // Listen for handshake response
            const unsubscribe = this.eventEmitter.on(HANDSHAKE_RESPONSE_EVENT, () => {
                clearTimeout(initTimeout);
                if (this.retryTimeout) {
                    clearTimeout(this.retryTimeout);
                    this.retryTimeout = null;
                }
                this.handshakeComplete = true;
                unsubscribe();
                resolve();
            });
            // Send handshake init message
            const handshakeMessage = {
                id: this.generateId(),
                type: 'handshake',
                event: HANDSHAKE_INIT_EVENT,
                handshakeType: 'request',
                sdkType: this.config.type,
                version: this.config.version,
                timestamp: Date.now(),
            };
            sendMessage(handshakeMessage);
        });
    }
    /**
     * Initializes the handshake process for the host side.
     * This method listens for handshake initialization messages from the client.
     */
    initializeHost() {
        // The host just needs to be ready to receive handshake messages
        // It doesn't proactively send anything
        console.debug('[host SDK] Host handshake manager initialized, waiting for client handshake...');
    }
    /**
     * Handles an incoming handshake message.
     * If it's an initialization message from the client, it will either respond or store it for later.
     * If it's a response message from the host, it will complete the handshake.
     *
     * @param message The received handshake message
     * @param sendResponse Function to send a response back (optional, if not provided will buffer the message)
     */
    handleHandshakeMessage(message, sendResponse) {
        if (message.type !== 'handshake') {
            return;
        }
        // Client received a handshake response from host
        if (this.config.type === 'client' && message.handshakeType === 'response') {
            console.debug('[client SDK] Received handshake response:', message);
            this.eventEmitter.emit(HANDSHAKE_RESPONSE_EVENT, message);
            this.handshakeComplete = true;
            return;
        }
        // Host received a handshake init from client
        if (this.config.type === 'host' && message.handshakeType === 'request' && message.event === HANDSHAKE_INIT_EVENT) {
            console.debug('[host SDK] Received handshake init:', message);
            if (sendResponse) {
                // We can respond immediately
                const responseMessage = {
                    id: message.id,
                    type: 'handshake',
                    event: HANDSHAKE_RESPONSE_EVENT,
                    handshakeType: 'response',
                    sdkType: this.config.type,
                    success: true,
                    version: this.config.version,
                    timestamp: Date.now(),
                };
                sendResponse(responseMessage);
                this.handshakeComplete = true;
            }
            else {
                // Store the message for later
                console.debug('[host SDK] Buffering handshake init message until iframe is ready');
                this.pendingInitMessage = message;
            }
        }
    }
    /**
     * Checks if there's a pending handshake init message that was received
     * before the target was ready.
     */
    hasPendingInitMessage() {
        return this.pendingInitMessage !== null;
    }
    /**
     * Processes any pending handshake init message that was received before
     * the target was ready.
     *
     * @param sendResponse Function to send a response
     * @returns True if a pending message was processed, false otherwise
     */
    processPendingInitMessage(sendResponse) {
        if (!this.pendingInitMessage) {
            return false;
        }
        console.debug('[host SDK] Processing buffered handshake init message');
        const message = this.pendingInitMessage;
        this.pendingInitMessage = null;
        const responseMessage = {
            id: message.id,
            type: 'handshake',
            event: HANDSHAKE_RESPONSE_EVENT,
            handshakeType: 'response',
            sdkType: this.config.type,
            success: true,
            version: this.config.version,
            timestamp: Date.now(),
        };
        sendResponse(responseMessage);
        this.handshakeComplete = true;
        return true;
    }
    /**
     * Checks if the handshake has been completed.
     *
     * @returns True if the handshake is complete, false otherwise.
     */
    isHandshakeComplete() {
        return this.handshakeComplete;
    }
    /**
     * Resets the handshake state, allowing for a new handshake.
     */
    reset() {
        this.handshakeComplete = false;
        this.retryCount = 0;
        this.pendingInitMessage = null;
        if (this.retryTimeout) {
            clearTimeout(this.retryTimeout);
            this.retryTimeout = null;
        }
    }
    /**
     * Generates a unique ID for handshake messages.
     *
     * @returns A unique ID string.
     */
    generateId() {
        return Math.random().toString(36).substring(2, 11);
    }
}

const byteToHex = [];
for (let i = 0; i < 256; ++i) {
    byteToHex.push((i + 0x100).toString(16).slice(1));
}
function unsafeStringify(arr, offset = 0) {
    return (byteToHex[arr[offset + 0]] +
        byteToHex[arr[offset + 1]] +
        byteToHex[arr[offset + 2]] +
        byteToHex[arr[offset + 3]] +
        '-' +
        byteToHex[arr[offset + 4]] +
        byteToHex[arr[offset + 5]] +
        '-' +
        byteToHex[arr[offset + 6]] +
        byteToHex[arr[offset + 7]] +
        '-' +
        byteToHex[arr[offset + 8]] +
        byteToHex[arr[offset + 9]] +
        '-' +
        byteToHex[arr[offset + 10]] +
        byteToHex[arr[offset + 11]] +
        byteToHex[arr[offset + 12]] +
        byteToHex[arr[offset + 13]] +
        byteToHex[arr[offset + 14]] +
        byteToHex[arr[offset + 15]]).toLowerCase();
}

let getRandomValues;
const rnds8 = new Uint8Array(16);
function rng() {
    if (!getRandomValues) {
        if (typeof crypto === 'undefined' || !crypto.getRandomValues) {
            throw new Error('crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported');
        }
        getRandomValues = crypto.getRandomValues.bind(crypto);
    }
    return getRandomValues(rnds8);
}

const randomUUID = typeof crypto !== 'undefined' && crypto.randomUUID && crypto.randomUUID.bind(crypto);
var native = { randomUUID };

function v4(options, buf, offset) {
    if (native.randomUUID && true && !options) {
        return native.randomUUID();
    }
    options = options || {};
    const rnds = options.random ?? options.rng?.() ?? rng();
    if (rnds.length < 16) {
        throw new Error('Random bytes length must be >= 16');
    }
    rnds[6] = (rnds[6] & 0x0f) | 0x40;
    rnds[8] = (rnds[8] & 0x3f) | 0x80;
    return unsafeStringify(rnds);
}

const AllowedOrigins = [
    'sitecore-staging.cloud',
    'sitecorecloud.app',
    'sitecorecloud.io'
];

const DEFAULT_TIMEOUT = 30000; // 30 seconds
/**
 * PostMessageBridge handles the communication between host and client applications
 * through the postMessage API. It provides:
 * - Type-safe message passing
 * - Request/response handling with timeouts
 * - Event emission and subscription
 * - Origin validation
 * - Handshake protocol for secure initialization
 */
class PostMessageBridge {
    config;
    eventEmitter = new EventEmitter();
    handshakeManager = null;
    /**
     * Stores pending requests with their resolve/reject handlers and timeout references.
     * The type parameter R represents the expected response type for each request.
     */
    pendingRequests = new Map();
    // Store request handlers for incoming requests.
    requestHandlers = new Map();
    // Store buffered messages received before the target was available
    bufferedMessages = [];
    initialized = false;
    source = 'sitecore-marketplace-sdk';
    connected = false;
    listenerMode = false;
    sessionId;
    sdkType;
    constructor(config) {
        this.config = config;
        window.addEventListener('message', this.handleMessage);
        // If no target is provided, we're in listener mode
        this.listenerMode = !config.target;
    }
    /**
     * Sets or updates the target window for postMessage communication.
     * This is useful for host applications that need to establish the iframe
     * target after the SDK is already initialized.
     *
     * @param target - The target window to communicate with (typically iframe.contentWindow)
     */
    setTarget(target) {
        if (!target) {
            throw this.formatError(new CoreError(exports.ErrorCode.INVALID_REQUEST, 'Target window cannot be null or undefined'));
        }
        this.config.target = target;
        this.listenerMode = false;
        // Process any pending handshake init messages
        if (this.handshakeManager && this.handshakeManager.hasPendingInitMessage()) {
            this.handshakeManager.processPendingInitMessage((message) => {
                this.postMessage(message);
                this.connected = true;
            });
        }
        // Process any buffered messages
        if (this.bufferedMessages.length > 0) {
            console.debug(`[${this.sdkType} PostMessageBridge] Processing ${this.bufferedMessages.length} buffered messages`);
            // Create a copy of the buffer to iterate over as we'll be clearing it
            const messagesToProcess = [...this.bufferedMessages];
            this.bufferedMessages = [];
            // Process each buffered message
            for (const { message, origin } of messagesToProcess) {
                this.processMessage(message, origin);
            }
        }
    }
    /**
     * Prefixes an error's message with the current SDK type (if available).
     * For example, errors will be transformed into "[client SDK] <original error message>".
     * @param error - The error to format.
     * @returns The updated error.
     */
    formatError(error) {
        const sdkTypePrefix = this.sdkType ? `[${this.sdkType} SDK] ` : '[SDK] ';
        error.message = sdkTypePrefix + error.message;
        return error;
    }
    /**
     * Initializes the bridge with basic configuration.
     * This only sets up the event listeners and basic configuration but doesn't perform the handshake yet.
     *
     * @param handshakeConfig - Configuration for the handshake
     */
    initialize(handshakeConfig) {
        if (this.initialized) {
            throw this.formatError(CoreError.alreadyInitialized());
        }
        this.sdkType = handshakeConfig.type;
        this.initialized = true;
        // Create the handshake manager
        this.handshakeManager = new HandshakeManager({
            type: handshakeConfig.type,
            targetOrigin: handshakeConfig.targetOrigin,
            selfOrigin: handshakeConfig.selfOrigin,
            version: handshakeConfig.version,
            timeout: this.config.timeout || DEFAULT_TIMEOUT,
        });
        // For the host, we just initialize the handshake manager
        if (handshakeConfig.type === 'host') {
            this.handshakeManager.initializeHost();
        }
    }
    /**
     * Connects the bridge by performing the handshake process.
     * For the client, this sends the handshake init message.
     * For the host, the connection is considered complete once a handshake init message is received and responded to.
     *
     * @returns A promise that resolves when the connection is established.
     */
    async connect() {
        if (!this.initialized) {
            throw this.formatError(CoreError.notInitialized());
        }
        if (this.connected) {
            return;
        }
        if (!this.handshakeManager) {
            throw this.formatError(new CoreError(exports.ErrorCode.INITIALIZATION_ERROR, 'Handshake manager not initialized'));
        }
        // For the client, we need to send the handshake init message
        if (this.sdkType === 'client') {
            if (this.listenerMode || !this.config.target) {
                throw this.formatError(new CoreError(exports.ErrorCode.INVALID_STATE, 'Cannot connect in listener mode without a target. Set target first.'));
            }
            try {
                await this.handshakeManager.initializeClient((message) => {
                    this.postMessage(message);
                });
                this.connected = true;
            }
            catch (error) {
                throw this.formatError(error);
            }
        }
        // For the host, the connection is considered complete when handshakeManager.isHandshakeComplete() is true
        // This happens after a handshake init message is received and responded to
        if (this.sdkType === 'host' && this.handshakeManager.isHandshakeComplete()) {
            this.connected = true;
        }
    }
    /**
     * Sends a request to the other side and waits for a response.
     * @param action - The action to perform
     * @param payload - Optional data to send with the request
     * @returns Promise that resolves with the response data
     */
    async request(action, payload) {
        this.ensureConnected();
        const id = this.generateId();
        const message = {
            id,
            type: 'request',
            action,
            payload,
            timestamp: Date.now(),
        };
        // Log outbound request details.
        console.debug(`[${this.sdkType} PostMessageBridge] Outbound Request - Action: ${action}, Message:`, message);
        return new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                if (this.pendingRequests.has(id)) {
                    this.pendingRequests.delete(id);
                    reject(this.formatError(CoreError.timeout(`Request for action "${action}" timed out after ${this.config.timeout || DEFAULT_TIMEOUT} ms.`)));
                }
            }, this.config.timeout || DEFAULT_TIMEOUT);
            this.pendingRequests.set(id, { resolve, reject, timeout });
            this.postMessage(message);
        });
    }
    /**
     * Emits an event to the other side.
     * @param event - Event name
     * @param payload - Data to send with the event
     */
    emit(event, payload) {
        this.ensureConnected();
        const message = {
            id: this.generateId(),
            type: 'event',
            event,
            payload,
            timestamp: Date.now(),
        };
        this.postMessage(message);
    }
    /**
     * Subscribes to events from the other side.
     * @param event - Event name to listen for
     * @param handler - Callback to handle the event data
     * @returns Function to unsubscribe from the event
     */
    on(event, handler) {
        return this.eventEmitter.on(event, handler);
    }
    /**
     * Cleans up all resources used by the bridge.
     * Should be called when the bridge is no longer needed.
     */
    destroy() {
        window.removeEventListener('message', this.handleMessage);
        this.eventEmitter.clear();
        this.pendingRequests.forEach(({ timeout }) => clearTimeout(timeout));
        this.pendingRequests.clear();
        this.bufferedMessages = [];
        this.handshakeManager = null;
        this.initialized = false;
        this.connected = false;
    }
    /**
     * Handles incoming postMessage events.
     * Validates origin and routes messages to appropriate handlers.
     */
    handleMessage = (event) => {
        const message = event.data;
        if (!message.source || message.source !== this.source) {
            // This ensures that the message is from the SDK and not from another source.
            return;
        }
        // Determine the expected origin.
        try {
            let isValidOrigin = this.isValidOrigin(event, message);
            if (!isValidOrigin) {
                const error = this.formatError(CoreError.invalidOrigin(event.origin));
                console.warn(error);
                // If the message comes from an unexpected origin, simply ignore it.
                return;
            }
        }
        catch (error) {
            console.warn(`[${this.sdkType} PostMessageBridge] Error parsing origin:`, error);
            return;
        }
        // If we're in listener mode (no target yet), buffer the message for later processing
        if (this.listenerMode || !this.config.target) {
            if (message.type === 'handshake') {
                // For handshake messages in listener mode, we can still handle them 
                // but can't respond until we have a target
                if (this.handshakeManager) {
                    this.handshakeManager.handleHandshakeMessage(message);
                }
            }
            // Buffer the message for processing once we have a target
            this.bufferedMessages.push({ message, origin: event.origin });
            console.debug(`[${this.sdkType} PostMessageBridge] Buffering message in listener mode:`, message);
            return;
        }
        // Process the message normally if we have a target
        this.processMessage(message, event.origin);
    };
    isValidOrigin(event, message) {
        let isValidOrigin = false;
        // For client SDK, we need to validate the origin during the handshake.
        if (this.sdkType === 'client' && message.type === 'handshake' && this.isNullOrEmpty(this.config.targetOrigin)) {
            isValidOrigin = AllowedOrigins.some(origin => event.origin.includes(origin));
            //Once origin is confirmed, targetOrigin will not be null
            if (isValidOrigin) {
                this.config.targetOrigin = event.origin;
            }
        }
        else {
            let expectedOrigin = this.config.targetOrigin;
            const eventOriginUrl = new URL(event.origin);
            const expectedOriginUrl = new URL(expectedOrigin);
            console.debug(`[${this.sdkType} PostMessageBridge] Received message:`, event.data, `from origin: ${eventOriginUrl.origin}. Expected origin: ${expectedOriginUrl.origin}`);
            isValidOrigin = eventOriginUrl.origin === expectedOriginUrl.origin;
        }
        return isValidOrigin;
    }
    /**
     * Processes a message after origin validation.
     *
     * @param message - The message to process
     * @param origin - The origin the message came from
     */
    processMessage(message, origin) {
        // Handle handshake messages
        if (message.type === 'handshake') {
            if (this.handshakeManager) {
                this.handshakeManager.handleHandshakeMessage(message, (responseMessage) => {
                    this.postMessage(responseMessage);
                    this.connected = true;
                });
            }
            return;
        }
        // Don't process other message types until connected
        if (!this.connected && this.sdkType === 'client') {
            console.warn(`[${this.sdkType} PostMessageBridge] Ignoring message, not connected yet:`, message);
            return;
        }
        switch (message.type) {
            case 'response':
                this.handleResponse(message);
                break;
            case 'event':
                this.handleEvent(message);
                break;
            case 'request':
                this.handleRequest(message);
                break;
        }
    }
    /**
     * Handles response messages by resolving their corresponding pending requests.
     */
    handleResponse(message) {
        // Log inbound response details.
        console.debug(`[${this.sdkType} PostMessageBridge] Handling response for message ID: %s`, message.id, message);
        const pending = this.pendingRequests.get(message.id);
        if (!pending)
            return;
        const { resolve, reject, timeout } = pending;
        clearTimeout(timeout);
        this.pendingRequests.delete(message.id);
        if (message.success) {
            resolve(message.data);
        }
        else {
            if (message.error) {
                reject(this.formatError(new CoreError(message.error.code, message.error.message, message.error.details)));
            }
            else {
                reject(this.formatError(new CoreError(exports.ErrorCode.UNKNOWN_ERROR, 'Unknown error')));
            }
        }
    }
    /**
     * Handles event messages by emitting them through the event emitter.
     */
    handleEvent(message) {
        this.eventEmitter.emit(message.event, message.payload);
    }
    /**
     * Handles incoming request messages by invoking the corresponding registered handler and sending back a response.
     * If no handler is registered, sends an error response.
     * @param message - The received request message.
     */
    async handleRequest(message) {
        console.debug(`[${this.sdkType} PostMessageBridge] Handling request for action: %s`, message.action, message);
        const fullKey = message.action.split(':');
        const actionWithoutPostfix = fullKey[0];
        const handler = this.requestHandlers.get(message.action) ?? this.requestHandlers.get(actionWithoutPostfix);
        const catchAllHandler = this.requestHandlers.get('*');
        if (!handler && !catchAllHandler) {
            // No handler registered for this action, send an error response.
            const response = {
                id: message.id,
                type: 'response',
                success: false,
                error: {
                    code: exports.ErrorCode.METHOD_NOT_FOUND,
                    message: `No handler found for request action "${message.action}".`,
                },
                timestamp: Date.now(),
            };
            this.postMessage(response);
            return;
        }
        try {
            let result;
            if (handler) {
                result = await handler(message.payload);
            }
            else if (catchAllHandler) {
                result = await catchAllHandler(message);
            }
            const response = {
                id: message.id,
                type: 'response',
                success: true,
                data: result,
                timestamp: Date.now(),
            };
            this.postMessage(response);
        }
        catch (err) {
            const response = {
                id: message.id,
                type: 'response',
                success: false,
                error: {
                    code: exports.ErrorCode.EXECUTION_ERROR,
                    message: err?.message || 'Error processing request',
                    details: err,
                },
                timestamp: Date.now(),
            };
            this.postMessage(response);
        }
    }
    /**
     * Sends a message to the other side through postMessage.
     */
    postMessage(message) {
        if (this.listenerMode || !this.config.target) {
            throw this.formatError(new CoreError(exports.ErrorCode.INVALID_STATE, 'Cannot send messages in listener mode. Set target first.'));
        }
        // Detailed logging added for tracing sent messages.
        message.source = this.source;
        console.debug(`[${this.sdkType} PostMessageBridge] Sending message:`, message, `to targetOrigin: ${this.config.targetOrigin}`);
        // Only client can be null for targetOrigin
        let targetOrigin;
        if (this.isNullOrEmpty(this.config.targetOrigin)) {
            targetOrigin = "*";
        }
        else {
            targetOrigin = this.config.targetOrigin;
        }
        this.config.target.postMessage(message, targetOrigin);
    }
    isNullOrEmpty(prop) {
        return prop == null || prop === "";
    }
    /**
     * Generates a unique message ID.
     */
    generateId() {
        return v4();
    }
    /**
     * Ensures the bridge is initialized and connected before allowing operations.
     * @throws {CoreError} If the bridge is not initialized or connected
     */
    ensureConnected() {
        if (!this.initialized) {
            throw this.formatError(CoreError.notInitialized());
        }
        if (this.listenerMode || !this.config.target) {
            throw this.formatError(new CoreError(exports.ErrorCode.INVALID_STATE, 'Cannot perform operations in listener mode. Set target first.'));
        }
        if (!this.connected && this.sdkType === 'client') {
            throw this.formatError(new CoreError(exports.ErrorCode.NOT_CONNECTED, 'Bridge is not connected. Call connect() first.'));
        }
    }
    /**
     * Registers a request handler for incoming requests.
     * @param action - The action to register the handler for
     * @param handler - The handler function to register
     */
    onRequest(action, handler) {
        this.requestHandlers.set(action, handler);
    }
    /**
     * Checks if the bridge is connected.
     * @returns True if the bridge is connected, false otherwise.
     */
    isConnected() {
        return this.connected;
    }
    /**
     * Checks if the bridge is in listener mode (no target set).
     * @returns True if in listener mode, false otherwise
     */
    isInListenerMode() {
        return this.listenerMode;
    }
}

// Re-export types and errors for consumers
/**
 * Core SDK for managing communication between host and client applications.
 * Provides a high-level API for bidirectional communication through postMessage.
 *
 * Features:
 * - Type-safe messaging
 * - Request/response pattern
 * - Event pub/sub system
 * - Secure handshake protocol
 * - Origin validation
 * // TODO: add examples
 */
class CoreSDK {
    bridge;
    /**
     * Creates a new instance of the Core SDK.
     * @param config - Configuration options for the SDK
     *
     * Note: If no target is provided in the config, the SDK will be initialized
     * in "listener mode" where it can receive messages but cannot send them until
     * a target is set using setTarget().
     */
    constructor(config) {
        this.bridge = new PostMessageBridge(config);
    }
    /**
     * Initializes the SDK with basic configuration. This method sets up event listeners
     * but doesn't perform the handshake yet. For the host, this allows listening for
     * handshake messages before the iframe is fully loaded.
     *
     * @param config - Handshake configuration
     * @throws {CoreError} If already initialized or handshake fails
     */
    initialize(config) {
        this.bridge.initialize(config);
    }
    /**
     * Sets or updates the target window for communication.
     * This is useful for host applications that need to establish the iframe
     * target after the SDK is already initialized.
     *
     * @param target - The target window to communicate with (typically iframe.contentWindow)
     */
    setTarget(target) {
        this.bridge.setTarget(target);
    }
    /**
     * Connects the SDK by performing the handshake with the other side.
     * For the client, this sends the handshake init message.
     * For the host, this completes the connection once a handshake init message is received and responded to.
     *
     * @returns Promise that resolves when the connection is established
     * @throws {CoreError} If already initialized
     */
    async connect() {
        return this.bridge.connect();
    }
    /**
     * Checks if the SDK is connected.
     * @returns True if connected, false otherwise
     */
    isConnected() {
        return this.bridge.isConnected();
    }
    /**
     * Checks if the SDK is in listener mode (no target set yet).
     * @returns True if in listener mode, false otherwise
     */
    isInListenerMode() {
        return this.bridge.isInListenerMode();
    }
    /**
     * Sends a request to the other side and waits for a response.
     * @param action - The action to perform
     * @param payload - Optional data to send with the request
     * @returns Promise that resolves with the response data
     * @throws {CoreError} If not initialized or request fails
     */
    async request(action, payload) {
        return this.bridge.request(action, payload);
    }
    /**
     * Emits an event to the other side.
     * @param event - Event name
     * @param payload - Data to send with the event
     * @throws {CoreError} If not initialized
     */
    emit(event, payload) {
        this.bridge.emit(event, payload);
    }
    /**
     * Subscribes to events from the other side.
     * @param event - Event name to listen for
     * @param handler - Callback to handle the event data
     * @returns Function to unsubscribe from the event
     */
    on(event, handler) {
        return this.bridge.on(event, handler);
    }
    /**
     * Registers a request handler for incoming requests.
     * @param action - The action to register the handler for
     * @param handler - The handler function to register
     */
    onRequest(action, handler) {
        this.bridge.onRequest(action, handler);
    }
    /**
     * Cleans up resources and removes event listeners.
     * The SDK instance cannot be used after calling this method.
     */
    destroy() {
        this.bridge.destroy();
    }
}

exports.CoreError = CoreError;
exports.CoreSDK = CoreSDK;
exports.GenericResponseData = GenericResponseData;
